### ISSUE TITLE
sporadic failure of wsn build: WsnConnectorTest.shouldNotHangOnToHttpConnectSessionsWhenEstablishingAndTearingDownWsnConnectorSessions

### ISSUE THREAD

**Issue by ahousing** [kaazing/gateway#162]
In one of 8 builds I did on my machine, I got the following failure.

```
WsnConnectorTest.shouldNotHangOnToHttpConnectSessionsWhenEstablishingAndTearingDownWsnConnectorSessions FAILED with exception org.junit.runners.model.TestTimedOutException: test timed out after 30000 milliseconds
org.junit.runners.model.TestTimedOutException: test timed out after 30000 milliseconds
	at java.lang.Object.wait(Native Method)
	at org.apache.mina.core.future.DefaultIoFuture.await(DefaultIoFuture.java:98)
	at org.apache.mina.core.future.DefaultConnectFuture.await(DefaultConnectFuture.java:106)
	at org.kaazing.gateway.transport.wsn.WsnConnectorTest.shouldNotHangOnToHttpConnectSessionsWhenEstablishingAndTearingDownWsnConnectorSessions(WsnConnectorTest.java:373)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:745)
=================== BEGIN STORED LOG MESSAGES ===========================
...
=================== END STORED LOG MESSAGES ===========================
shouldCloseWsnSessionWhenTransportClosesCleanlyButUnexpectedly(org.kaazing.gateway.transport.wsn.WsnConnectorTest) starting
WsnConnectorTest.shouldCloseWsnSessionWhenTransportClosesCleanlyButUnexpectedly success
shouldCorrectlyConstructLocalAndRemoteAddressesForConnectedWsnSessions(org.kaazing.gateway.transport.wsn.WsnConnectorTest) starting
WsnConnectorTest.shouldCorrectlyConstructLocalAndRemoteAddressesForConnectedWsnSessions success
Tests run: 3, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 30.229 sec <<< FAILURE!
Running org.kaazing.gateway.transport.wsn.WsnTransportTest
connectorShouldWriteAndReceiveMessage(org.kaazing.gateway.transport.wsn.WsnTransportTest) starting
Acceptor: received message: Message from connector
WsnTransportTest.connectorShouldWriteAndReceiveMessage success
connectorShouldReceiveMessageFromAcceptor(org.kaazing.gateway.transport.wsn.WsnTransportTest) starting
WsnTransportTest.connectorShouldReceiveMessageFromAcceptor success
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.022 sec

Results :

Tests in error: 
  shouldNotHangOnToHttpConnectSessionsWhenEstablishingAndTearingDownWsnConnectorSessions(org.kaazing.gateway.transport.wsn.WsnConnectorTest): test timed out after 30000 milliseconds
```

**Comment by ahousing** [comment]
The snipped out part (the "stored log messages" is in: 
[zenhub.log | uploaded via ZenHub](https://files.zenhub.io/55677f22cfc3d8324e9aaee0)

**Comment by jitsni** [comment]
From the logs, it only completed 92 iterations out of 100 in the specified time. So reducing the no of iterations in the test. Don't know whether it fixes.

Created a pull request: 
https://github.com/kaazing/gateway/pull/175

**Comment by jitsni** [comment]
The WsnConnector#connect method is called in junit thread in the test.

When the test passes, the order is as follows:
1. HttpConnector ConnectFuture listener gets called in i/o thread
2. HttpSession is closed in i/o thread

When the test fails, the order is as follows:
1. HttpSession is closed in i/o thread
2. HttpConnector ConnectFuture listener gets called in junit thread

---------------------- Failure ---------
11179 [New I/O worker #434] INFO  connect  - HttpConnector Got INFO_SWITCHING_PROTOCOLS
11179 [New I/O worker #434] INFO  HttpConnectProcessor  - HttpConnectProcessor#removeInternal with upgradeHandler=null
...
11180 [Time-limited test] INFO  connect  - Setting ioBridgeHandler for upgrade
----------------------------

**Comment by jitsni** [comment]
junit test thread does the following in order:
(1) WsnConnector.connect()
(2) HttpConnector.connect()
(3) Adds a listener to Http's ConnectFuture that sets up upgrade handler

Between (2) and (3), an I/O thread connects to  server, gets a response 101, and closes HttpSession. HttpConnectProcessor tries to upgrade, but doesn't find upgrade handler as (3) is not yet done.

Perhaps, this race happens extremely low in practice as there is http request/response network roundtrip.

**Comment by sbadugu** [comment]
There was another sporadic [build failure](https://s3.amazonaws.com/archive.travis-ci.org/jobs/67591431/log.txt) with same message.

**Comment by sbadugu** [comment]
This happened [again](https://s3.amazonaws.com/archive.travis-ci.org/jobs/68992591/log.txt). Probably, we should fix this sooner than later.

**Comment by sbadugu** [comment]
@jitsni Can we prioritize this issue and get it resolved if you already know what the problem is ?

**Comment by irina-mitrea-luxoft** [comment]
Attempted to modify timeout.
Reproduced the issue on my local build intermittently, attaching the logs with failure/success scenarios.

[failed_to_get_session.txt | uploaded via ZenHub](https://files.zenhub.io/55f14d8a5dbd8ea97b0bc975)


[timeout.txt | uploaded via ZenHub](https://files.zenhub.io/55f14db35dbd8ea97b0bc977)

[successful.txt | uploaded via ZenHub](https://files.zenhub.io/55f14db85dbd8ea97b0bc978)

**Comment by irina-mitrea-luxoft** [comment]
Attempted to increase timeout interval.
Reproduced the issue on my local build intermittently, attaching logs with several failure/success situations.


[failed_to_get_session.txt | uploaded via ZenHub](https://files.zenhub.io/55f1529d5dbd8ea97b0bc9ab)

[successful.txt | uploaded via ZenHub](https://files.zenhub.io/55f152a35dbd8ea97b0bc9ac)

[timeout.txt | uploaded via ZenHub](https://files.zenhub.io/55f152a75dbd8ea97b0bc9ad)

**Comment by dpwspoon** [comment]
Hoping this is resolved via #328, we can reopen if reproduces
